python/cuml/dask/common/dask_arr_utils.py

```diff
@@ -160,49 +109,51 @@ def to_sp_dask_array(cudf_or_array, client=None):
 
     meta = cupyx.scipy.sparse.csr_matrix(rmm_cupy_ary(cp.zeros, 1))
 
-    if isinstance(cudf_or_array, dask.array.Array):
+    ret = cudf_or_array
+
+    if isinstance(ret, dask.array.Array):
         # At the time of developing this, using map_blocks will not work
         # to convert a Dask.Array to CuPy sparse arrays underneath.
 
-        parts = client.sync(_extract_partitions, cudf_or_array)
-        cudf_or_array = [client.submit(_conv_np_to_df, part, workers=[w])
-                         for w, part in parts]
+        parts = client.sync(_extract_partitions, ret)
+        futures = [client.submit(_conv_np_to_df, part, workers=[w], pure=False)
+                   for w, part in parts]
 
-        cudf_or_array = to_dask_cudf(cudf_or_array)
+        ret = to_dask_cudf(futures)
 
-    if isinstance(cudf_or_array, dask.dataframe.DataFrame):
+    if isinstance(ret, dask.dataframe.DataFrame):
         """
         Dask.Dataframe needs special attention since it has multiple dtypes.
         Just use the first (and assume all the rest are the same)
         """
-        cudf_or_array = cudf_or_array.map_partitions(
+        ret = ret.map_partitions(
             _conv_df_to_sp, meta=dask.array.from_array(meta))
 
         # This will also handle the input of dask.array.Array
-        return cudf_or_array
+        return ret
 
     else:
```

Comment:
The logic for this function has gotten pretty long and the control flow is a little confusing. It's a lot of conditionals, some of which return and some of which do not, with some that are nested. Can you lift the "if dataframe" higher in the function and make it clear that that one is an early return, then un-nest the big nested block here (with if scipy.sparse.issppmatrix)? would be nice to ensure then that none of these does an early return except for the dataframe one... I think that would help.

Reply:
This function is a series of workarounds for the fact that we cannot yet directly create a Cupy-backed sparse dask array. 

The general control flow can be read as:
1) If a dense (numpy) dask array is passed in, convert it to a dask cudf. 
2) If we have a dask cudf, perform a map_partitions to convert it to an array and return it. 
3) Otherwise, what we've received is not a dask collection, so it's either already a sparse matrix, or it's a cudf, or a numpy dense array, etc... @divyegala, we had discussed this previously, but that last return statement is part of the else so it should be tabbed over. The returns are not early- they result from the if..else. I agree the branching logic could be cleaned up by putting the if..elif..elif..etc... branches inside the initial else clause inside it's own function- maybe something like `_convert_to_sparse_array`. @divyegala, if it's okay with you, I'll clean this one up.

Reply:
Sounds good @cjnolet. Let me know when done and I can address the other stuff.

Reply:
Just pushed a cleaned up version.